Documentation/proposals/202411-zone-aware-sharding.md

```diff
@@ -0,0 +1,321 @@
+
+All values used in this snippet should also be the defaults for their
+corresponding keys.
+
+```yaml
+spec:
+  shardingStrategy:
+     # Select a sharding mode. Can be 'Classic' or 'Topology'.
+     # Defaults to `Classic`.
+    mode: 'Classic'    
+
+    # The following section is only valid if "mode" is set to "Classic"
+    classic:
+        # Metric label used for sharding. Defaults to `__address__`.
+        sourceLabel: '__address__'
+      
+    # The following section is only valid if "mode" is set to "Topology"
+    # 'topology.kubernetes.io/zone' for 'topology'
+    topology: 
+        # A kubernetes node label defining the topology to be sharded on.
+        # Defaults to `topology.kubernetes.io/zone`
+        nodeLabel: 'topology.kubernetes.io/zone'
+        
+        # A prometheus metric containing the topology value of a given target.
+        # Defaults to `__meta_kubernetes_node_label_topology_kubernetes_io_zone`
+        sourceLabel: '__meta_kubernetes_node_label_topology_kubernetes_io_zone'
+
+        # All topology values to be used by nodeLabel and sourceLabel
+        values: []
+```
+
+The `additionalRelabelConfig` section is meant to allow the `sourceLabel` to be
+generated if needed. This should allow enough flexibility to cover edgecases
+not anticiapted by this proposal.
+
+It is also to be noted that the `topology` section does not use the term `zone`.
+This makes the feature more flexible in case a user needs to shard on e.g.
+regions instead.
+
+## Generated configuration
+
+The following examples are based on the algorithm above.
+Please note that `shard_index` has to be provided by the operator during
+config generation.
+
+We use a replica count of 2 in all examples to illustrate that this value
+does not have any effect, as both replicas will have the same `shared_index`
+assigned.
+
+### classic mode
+
+Given the following configuration
+
+```yaml
+spec:
+  shards: 4
+  replicas: 2
+  shardingStrategy:
+    mode: 'Classic'    
+    classic:
+      sourceLabel: '__address__'
+```
+
+we would get the following output for `shard_index == 2`
+
+```yaml
+- source_labels: 
+    # shardingStrategy.classic.sourceLabel
+    - '__address__'
+  modulus: 4                    # number of shards
+  target_label: '__tmp_hash'
+  action: 'hashmod'
+- source_labels: ['__tmp_hash']
+  regex: '2'                    # shard_index
+  action: 'keep'
+```
+
+### topology mode
+
+Given the following configuration
+
+```yaml
+spec:
+  shards: 4
+  replicas: 2
+  shardingStrategy:
+    mode: 'Topology'    
+    topology:
+      nodeLabel: 'topology.kubernetes.io/zone'
+      sourceLabel: '__meta_kubernetes_node_label_topology_kubernetes_io_zone'
+      values:
+        - europe-west4-a
+        - europe-west4-b
+```
+
+we would get the following output for `shard_index == 2`:
+
+```yaml
```

Comment:
Sure, but I currently cannot see/remember where this comes from 😅

Reply:
Is this what you meant?

```yaml
spec:
  scrapeClasses:
    - name: 'topology'
      default: true
      attachMetadata:
        node: true
```

Reply:
No I meant to add it to the generated Prometheus configuration but don't add it now (it will be obvious when implementing).

````
- job_name: serviceMonitor/foo/bar/0
  honor_labels: false
  kubernetes_sd_configs:                                                                       
  - role: endpoints                                                                            
    attach_metadata: <--- here
      node: true
    namespaces:    
      names:       
      - foo                                                                                
    relabel_configs:
    - ...

````

Reply:
I added this as a note so it is documented.